tests/test_portfolio_env.py

```diff
@@ -0,0 +1,282 @@
+        datums.add().rows(*series)
+
+
+def test_reset_returns_first_value_from_datums(make_env, datums):
+    datums.add().rows([1], [2])
+    assert_obs_eq(make_env().reset(), [1])
+
+
+@pytest.mark.parametrize('num_values, assets, window_size', [
+    (3, 1, 1),
+    (3, 2, 1),
+    (1, 3, 1),
+    (3, 1, 2),
+    (1, 1, 5),
+])
+def test_observations_follow_observation_space_shape(make_env, datums, num_values, assets, window_size):
+    fill_datums(datums, assets, num_values)
+    env = make_env(window_size)
+    assert env.reset().shape == env.observation_space.shape
+
+
+def test_reset_returns_properly_windowed_observation(make_env, datums):
+    datums.add().rows([1], [2], [3])
+    assert_obs_eq(make_env(window_size=2).reset(), [[1, 2]])
+
+
+def test_reset_combined_setup_observation(make_env, datums):
+    datums.add().rows(
+        [2, 1, 3],
+        [3, 2, 4],
+        [1, 1, 1],
+    )
+    datums.add().rows(
+        [1e-6, 1e-9, 1],
+        [1, 1e-6, 2],
+        [1, 1, 1],
+    )
+    assert_obs_eq(make_env(window_size=2).reset(), [[[2, 3], [1e-6, 1]],
+                                                    [[1, 2], [1e-9, 1e-6]],
+                                                    [[3, 4], [1, 2]]])
+
+
+def test_calculate_returns_as_observations_when_configured(make_env, datums):
+    datums.add().rows(
+        [2, 1, 4],
+        [4, 2, 8],
+        [1, 4, 2],
+        [1, 1, 1],
+    )
+    assert_obs_eq(make_env(window_size=2, calc_returns=True).reset(), [[2, 0.25],
+                                                                       [2, 2],
+                                                                       [2, 0.25]])
+
+
+def test_stepping_the_environment_returns_next_observation(make_ready_env, datums):
+    datums.add().rows([1], [2], [3])
+    env = make_ready_env()
+    assert_obs_eq(unpack_obs(idle_step(env)), [2])
+    assert_obs_eq(unpack_obs(idle_step(env)), [3])
+
+
+def idle_step(env):
+    a = normalize(env.portfolio.assets)
+    return env.step(a)
+
+
+def test_resetting_the_environment_resets_observations(make_ready_env, datums):
+    datums.add().rows([1], [2], [3])
+    env = make_ready_env()
+    assert_obs_eq(unpack_obs(idle_step(env)), [2])
+    env.reset()
+    assert_obs_eq(unpack_obs(idle_step(env)), [2])
+
+
+def test_environment_is_done_at_the_last_observation(make_ready_env, datums):
+    datums.add().rows([1], [2], [3])
+    env = make_ready_env()
+    assert not unpack_done(idle_step(env))
+    assert unpack_done(idle_step(env))
+
+
+def test_raise_an_error_when_not_enough_data_for_a_single_step(make_env, datums):
+    datums.add().rows([1], [2])
+    env = make_env(window_size=2)
+    with pytest.raises(DatumsError):
+        env.reset()
+
+
+def test_raise_an_error_when_stepping_past_done(make_ready_env, datums):
+    datums.add().rows([1], [2])
+    env = make_ready_env()
+    list(until_done(env, [1, 0]))
+    with pytest.raises(DatumsError):
```

Comment:
Is there maybe a special gym error for stepping into done environments? Possibly inherit from that